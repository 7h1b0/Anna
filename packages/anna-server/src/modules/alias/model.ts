import Ajv from 'ajv';
import { v4 as uuidv4 } from 'uuid';
import knex from '../../knexClient';
import aliasSchema from './schema';
import { omit } from '../../utils/utils';

export const TABLE = 'alias';
export const COLUMNS = [
  'aliasId',
  'sceneId',
  'name',
  'description',
  'enabled',
  'startTime',
  'endTime',
  'createdAt',
  'updatedAt',
  'createdBy',
];

type Alias = {
  aliasId: string;
  sceneId: string;
  name: string;
  description?: string;
  startTime?: number;
  endTime?: number;
  enabled: boolean;
  createdAt: number;
  updatedAt: number;
  createdBy: string;
};

export function validate(data: Record<string, unknown>) {
  const ajv = new Ajv();
  return ajv.validate(aliasSchema, data);
}

export async function findAll(): Promise<Alias[]> {
  return knex(TABLE).select(COLUMNS);
}

export async function findById(aliasId: string): Promise<Alias> {
  return knex(TABLE).first(COLUMNS).where('aliasId', aliasId);
}

export async function findByName(name: string): Promise<Alias[]> {
  return knex(TABLE).select(COLUMNS).where('name', name);
}

export async function save({
  sceneId,
  name,
  description,
  createdBy,
  enabled = true,
  startTime,
  endTime,
}: {
  sceneId: string;
  name: string;
  description: string;
  createdBy: string;
  enabled?: boolean;
  startTime?: number;
  endTime?: number;
}): Promise<string> {
  const aliasId = uuidv4();
  await knex(TABLE).insert({
    aliasId,
    sceneId,
    description,
    name,
    enabled,
    createdBy,
    startTime,
    endTime,
  });
  return aliasId;
}

export function remove(aliasId: string) {
  return knex(TABLE).where('aliasId', aliasId).del();
}

export function findByIdAndUpdate(aliasId: string, payload: Partial<Alias>) {
  const safePayload = omit(
    payload,
    'aliasId',
    'createdAt',
    'updatedAt',
    'createdBy',
  );
  return knex(TABLE).update(safePayload).where('aliasId', aliasId);
}

export function isInteger(value: unknown | number): value is number {
  return Number.isInteger(value);
}

/**
 * Given aliases, returns one alias respecting the constraints
 */
export function resolveActiveAlias(aliases: Alias[]): Alias | undefined {
  const time = new Date().getHours();
  const activeAlias = aliases
    .filter((alias) => alias.enabled)
    .filter((alias) => {
      if (isInteger(alias.startTime) && isInteger(alias.endTime)) {
        if (alias.startTime < alias.endTime) {
          return time >= alias.startTime && time < alias.endTime;
        }
        return time >= alias.startTime || time < alias.endTime;
      }
      return true;
    });

  return activeAlias.pop();
}
