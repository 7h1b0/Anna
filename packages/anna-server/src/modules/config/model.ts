import knex from '../../knexClient';

export const TABLE = 'config';

export type Config = {
  isAway: boolean;
};

export async function fetchIsAway(): Promise<{ isAway: true }> {
  return knex(TABLE).first('isAway');
}

export function updateIsAway(isAway: boolean) {
  return knex(TABLE).update({ isAway });
}
