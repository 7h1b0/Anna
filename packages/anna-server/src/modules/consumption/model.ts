import knex from '../../knexClient';

export const TABLE = 'consumption';
export const COLUMNS = ['date', 'value'];

export type Consumption = {
  date: number;
  value: number;
};

export async function findLastEntries(limit: number): Promise<Consumption[]> {
  return knex(TABLE)
    .select(COLUMNS)
    .whereNotNull('value')
    .orderBy('date', 'desc')
    .limit(limit);
}

export async function findLastEntry(): Promise<Consumption> {
  return knex(TABLE)
    .first(COLUMNS)
    .whereNotNull('value')
    .orderBy('date', 'desc')
    .limit(1);
}

export async function save(consumptions: { date: Date; value: number }[]) {
  return knex.transaction((trx) => {
    return Promise.all(
      consumptions.map((consumption) => trx.insert(consumption).into(TABLE)),
    );
  });
}
