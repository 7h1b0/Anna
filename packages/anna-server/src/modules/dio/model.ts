import Ajv from 'ajv';
import knex from '../../knexClient';
import dioSchema from './schema';

export const TABLE = 'dios';

export type Dio = {
  dioId: number;
  roomId: string;
  name: string;
};

export function validate(data: Record<string, unknown>) {
  const ajv = new Ajv();
  return ajv.validate(dioSchema, data);
}

export async function findAll(): Promise<Dio[]> {
  return knex(TABLE).select('dioId', 'roomId', 'name');
}

export async function findById(dioId: number): Promise<Dio> {
  return knex(TABLE).first('dioId', 'roomId', 'name').where('dioId', dioId);
}

export async function findByRoomId(roomId: string): Promise<Dio[]> {
  return knex(TABLE).select('dioId', 'roomId', 'name').where('roomId', roomId);
}

export async function save(newDio: Dio): Promise<Dio> {
  return knex(TABLE)
    .insert(newDio)
    .then(() => newDio);
}

export function remove(dioId: number) {
  return knex(TABLE).where('dioId', dioId).del();
}

export function findByIdAndUpdate(
  dioId: number,
  payload: Partial<Omit<Dio, 'dioId'>>,
) {
  return knex(TABLE).update(payload).where('dioId', dioId);
}
