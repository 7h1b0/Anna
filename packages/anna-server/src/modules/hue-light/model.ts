import knex from '../../knexClient';
export const TABLE = 'lights';
export const COLUMNS = ['roomId', 'lightId'];

export function findRoomId(lightId: number): Promise<string[]> {
  return knex(TABLE).first('roomId').where('lightId', lightId);
}

export function findAll() {
  return knex(TABLE).select(...COLUMNS);
}

export function findByRoomId(roomId: string): Promise<number[]> {
  return knex(TABLE)
    .select('lightId')
    .where('roomId', roomId)
    .then((res) => res.map(({ lightId }) => lightId));
}

export function remove(lightId: number): Promise<number> {
  return knex(TABLE).where('lightId', lightId).del();
}

export async function insertOrUpdate(lightId: number, roomId: string) {
  await knex.raw(
    'INSERT INTO lights (lightId, roomId) values (?, ?) ON DUPLICATE KEY UPDATE roomId=VALUES(roomId)',
    [lightId, roomId],
  );
}
