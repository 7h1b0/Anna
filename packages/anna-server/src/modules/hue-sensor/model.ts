import Ajv from 'ajv';
import knex from '../../knexClient';
import sensorSchema from './schema';

export const TABLE = 'sensors';
export const COLUMNS = ['roomId', 'sensorId'];

export function findByRoomId(roomId: string): Promise<string[]> {
  return knex(TABLE)
    .select('sensorId')
    .where('roomId', roomId)
    .then((res) => res.map(({ sensorId }) => sensorId));
}

export function findAll() {
  return knex(TABLE).select(...COLUMNS);
}

export function remove(sensorId: string): Promise<number> {
  return knex(TABLE).where('sensorId', sensorId).del();
}

export async function insertOrUpdate(sensorId: string, roomId: string) {
  await knex.raw(
    'INSERT INTO sensors (sensorId, roomId) values (?, ?) ON DUPLICATE KEY UPDATE roomId=VALUES(roomId)',
    [sensorId, roomId],
  );
}

export function validate(data: object) {
  const ajv = new Ajv();
  return ajv.validate(sensorSchema, data);
}
