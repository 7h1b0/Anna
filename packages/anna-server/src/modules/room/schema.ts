export default {
  $id: 'room',
  type: 'object',
  additionalProperties: false,
  properties: {
    roomId: {
      type: 'string',
      pattern: '[a-z0-9-]{36}',
    },
    name: {
      type: 'string',
      minLength: 3,
    },
    description: {
      type: 'string',
      minLength: 3,
    },
  },
  required: ['name'],
};
