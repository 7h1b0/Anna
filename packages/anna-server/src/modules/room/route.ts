import { Router } from 'express';
import * as Room from '../../modules/room/model';
import { findByRoomId as findDio } from '../../modules/dio/model';
import { getLights, getSensorsByRoomId } from '../../services/hueService';

const routes = Router();
routes
  .route('/api/rooms')
  .get((req, res) => {
    Room.findAll()
      .then((rooms) => res.json(rooms))
      .catch((err) => res.status(500).send({ err }));
  })
  .post((req, res) => {
    const isValid = Room.validate(req.body);
    if (!isValid) {
      res.sendStatus(400);
    } else {
      const userId = res.locals.user.userId;
      Room.save({ ...req.body, userId })
        .then((newRoomId) => res.status(201).json(newRoomId))
        .catch((err) => res.status(500).send({ err }));
    }
  });

routes
  .route('/api/rooms/:room_id([a-fA-F0-9-]{36})')
  .get((req, res) => {
    const roomId = req.params.room_id;
    const fetchDio = findDio(roomId);
    const fetchAllLights = getLights();
    const fetchSensors = getSensorsByRoomId(roomId);
    const fetchRoom = Room.findById(roomId);

    Promise.all([fetchRoom, fetchDio, fetchAllLights, fetchSensors])
      .then(([room, dios, lights, sensors]) => {
        if (!room) {
          return res.sendStatus(404);
        }
        const filteredLights = lights.filter(
          (light) => light.roomId === roomId,
        );

        res.json(
          Object.assign(room, {
            devices: {
              dios,
              hueLights: filteredLights,
            },
            sensors,
          }),
        );
      })
      .catch((err) => {
        res.status(500).send({ err });
      });
  })
  .patch((req, res) => {
    const isValid = Room.validate(req.body);
    if (!isValid) {
      res.sendStatus(400);
    } else {
      Room.findByIdAndUpdate(req.params.room_id, req.body)
        .then((rowsAffected) => {
          if (rowsAffected < 1) {
            res.sendStatus(404);
          } else {
            res.sendStatus(204);
          }
        })
        .catch((err) => res.status(500).send({ err }));
    }
  })
  .delete((req, res) => {
    Room.remove(req.params.room_id)
      .then((rowsAffected) => {
        if (rowsAffected < 1) {
          res.sendStatus(403);
        } else {
          res.sendStatus(204);
        }
      })
      .catch((err) => res.status(500).send({ err }));
  });

export default routes;
