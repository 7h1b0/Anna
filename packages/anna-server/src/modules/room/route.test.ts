import request from 'supertest';
import { createUser } from 'factories';
import knex from '../../knexClient';
import * as Room from '../../modules/room/model';
import * as User from '../../modules/user/model';
import * as Dio from '../../modules/dio/model';
import app from '../../index';
import * as hueService from '../../services/hueService';

const user = createUser({ userId: 'c10c80e8-49e4-4d6b-b966-4fc9fb98879f' });
const initRooms = [
  {
    roomId: '0fc1d78e-fd1c-4717-b610-65d2fa3d01b2',
    description: 'this is a test',
    name: 'room_1',
    createdBy: 'c10c80e8-49e4-4d6b-b966-4fc9fb98879f',
    createdAt: new Date('2018-01-01'),
    updatedAt: new Date('2018-01-02'),
  },
  {
    roomId: 'c10c80e8-49e4-4d6b-b966-4fc9fb98879f',
    description: 'this is a second test',
    name: 'room_2',
    createdBy: 'c10c80e8-49e4-4d6b-b966-4fc9fb98879f',
    createdAt: new Date('2018-01-01'),
    updatedAt: new Date('2018-01-02'),
  },
  {
    roomId: 'd10c80e8-49e4-4d6b-b966-4fc9fb98879f',
    description: 'this is a second test',
    name: 'room_3',
    createdBy: 'c10c80e8-49e4-4d6b-b966-4fc9fb98879f',
    createdAt: new Date('2018-01-01'),
    updatedAt: new Date('2018-01-02'),
  },
];
const initDios = [
  {
    dioId: 1,
    roomId: initRooms[0].roomId,
    name: 'test_1',
  },
  {
    dioId: 2,
    roomId: initRooms[0].roomId,
    name: 'test_2',
  },
  {
    dioId: 3,
    roomId: initRooms[1].roomId,
    name: 'test_3',
  },
];

describe('Rooms API', () => {
  beforeAll(async () => {
    await knex(User.TABLE).truncate();
    await knex(User.TABLE).insert(user);

    await knex(Dio.TABLE).truncate();
    await knex(Dio.TABLE).insert(initDios);
  });

  beforeEach(async () => {
    await knex(Room.TABLE).truncate();
    await knex(Room.TABLE).insert(initRooms);
  });

  describe('/api/rooms', () => {
    describe('GET', () => {
      it('should return all rooms', async () => {
        const response = await request(app)
          .get('/api/rooms')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.body).toMatchSnapshot();
      });

      it('should return 401 when user is not authenticated', async () => {
        const response = await request(app)
          .get('/api/rooms')
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBeUnauthorized();
      });
    });

    describe('POST', () => {
      it('should create a new room', async () => {
        const response = await request(app)
          .post('/api/rooms')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send({
            name: 'Living Room',
            description: 'a description',
          });

        expect(response.status).toHaveStatusOk();

        const rooms = await knex(Room.TABLE)
          .first()
          .where('roomId', response.body);

        expect(rooms).toMatchSnapshot({
          roomId: expect.stringMatching(/[a-fA-F0-9-]{36}/),
          createdAt: expect.any(Number),
          updatedAt: expect.any(Number),
        });
      });

      it('should return 401 when user is not authenticated', async () => {
        const response = await request(app)
          .post('/api/rooms')
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBeUnauthorized();
      });

      it('should retun 400 if request is invalid', async () => {
        const response = await request(app)
          .post('/api/rooms')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send({
            description: 'missing name',
          });

        expect(response.status).toBeBadRequest();
      });
    });
  });

  describe('/api/rooms/:id', () => {
    let mockGetLights: jest.SpyInstance<Promise<hueService.HueLight[]>, []>;
    let mockGetSensors: jest.SpyInstance<
      Promise<hueService.TemperatureSensor[]>,
      [string]
    >;
    beforeAll(() => {
      mockGetLights = jest.spyOn(hueService, 'getLights').mockResolvedValue([
        {
          id: '1',
          roomId: initRooms[0].roomId,
          name: 'light_1',
          type: 'Dimmable light',
          state: {
            on: true,
            bri: 1,
            reachable: true,
          },
        },
        {
          id: '2',
          roomId: initRooms[0].roomId,
          name: 'light_2',
          type: 'Dimmable light',
          state: {
            on: true,
            bri: 1,
            reachable: true,
          },
        },
        {
          id: '3',
          roomId: initRooms[1].roomId,
          name: 'light_3',
          type: 'Dimmable light',
          state: {
            on: true,
            bri: 1,
            reachable: true,
          },
        },
      ]);
      mockGetSensors = jest
        .spyOn(hueService, 'getSensorsByRoomId')
        .mockResolvedValue([
          {
            id: '10',
            state: {
              temperature: 20,
              lastupdated: '2020-05-09T19:10:00',
            },
            type: 'ZLLTemperature',
          },
        ]);
    });

    afterEach(() => {
      mockGetLights.mockClear();
      mockGetSensors.mockClear();
    });

    describe('GET', () => {
      it('should return an room', async () => {
        const response = await request(app)
          .get(`/api/rooms/${initRooms[0].roomId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.body).toMatchSnapshot();
      });

      it('should return 401 when user is not authenticated', async () => {
        const response = await request(app)
          .get(`/api/rooms/${initRooms[0].roomId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBeUnauthorized();
      });

      it("should retun 404 when a room don't exist", async () => {
        const response = await request(app)
          .get('/api/rooms/c10c80e8-49e4-4d6b-b966-4fc9fb9887aa')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.status).toBe(404);
      });
    });

    describe('PATCH', () => {
      it('should update an room', async () => {
        const response = await request(app)
          .patch(`/api/rooms/${initRooms[0].roomId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send({
            description: 'this is a second test',
            name: 'room_updated',
          });

        expect(response.status).toHaveStatusOk();

        const room = await knex(Room.TABLE)
          .first()
          .where('roomId', initRooms[0].roomId);

        expect(room).toMatchSnapshot({
          roomId: expect.stringMatching(/[a-fA-F0-9-]{36}/),
          createdAt: expect.any(Number),
          updatedAt: expect.any(Number),
        });
      });

      it('should return 401 when user is not authenticated', async () => {
        const response = await request(app)
          .patch(`/api/rooms/${initRooms[0].roomId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBeUnauthorized();
      });

      it('should return 400 when request is invalid', async () => {
        const response = await request(app)
          .patch(`/api/rooms/${initRooms[0].roomId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send({
            name: 'test_update',
            fake: true,
          });

        expect(response.status).toBeBadRequest();
      });

      it("should retun 404 when a room don't exist", async () => {
        const response = await request(app)
          .patch('/api/rooms/c10c80e8-49e4-4d6b-b966-4fc9fb9887aa')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send({
            description: 'this is a second test',
            name: 'room_updated',
          });

        expect(response.status).toBe(404);
      });
    });

    describe('DELETE', () => {
      it('should delete an room', async () => {
        const response = await request(app)
          .delete(`/api/rooms/${initRooms[2].roomId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.status).toHaveStatusOk();

        const room = await knex(Room.TABLE)
          .select()
          .where('roomId', initRooms[2].roomId);

        expect(room).toHaveLength(0);
      });

      it('should return 403 when room is used', async () => {
        const response = await request(app)
          .delete(`/api/rooms/${initRooms[0].roomId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.status).toBe(403);
      });

      it('should return 401 when user is not authenticated', async () => {
        const response = await request(app)
          .delete(`/api/rooms/${initRooms[0].roomId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBeUnauthorized();
      });

      it("should retun 403 when a room don't exist", async () => {
        const response = await request(app)
          .delete('/api/rooms/c10c80e8-49e4-4d6b-b966-4fc9fb9887aa')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.status).toBe(403);
      });
    });
  });
});
