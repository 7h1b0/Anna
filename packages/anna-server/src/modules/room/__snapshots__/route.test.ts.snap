// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rooms API /api/rooms GET should return all rooms 1`] = `
Array [
  Object {
    "createdAt": 1514764800000,
    "createdBy": "c10c80e8-49e4-4d6b-b966-4fc9fb98879f",
    "description": "this is a test",
    "name": "room_1",
    "roomId": "0fc1d78e-fd1c-4717-b610-65d2fa3d01b2",
    "updatedAt": 1514851200000,
  },
  Object {
    "createdAt": 1514764800000,
    "createdBy": "c10c80e8-49e4-4d6b-b966-4fc9fb98879f",
    "description": "this is a second test",
    "name": "room_2",
    "roomId": "c10c80e8-49e4-4d6b-b966-4fc9fb98879f",
    "updatedAt": 1514851200000,
  },
  Object {
    "createdAt": 1514764800000,
    "createdBy": "c10c80e8-49e4-4d6b-b966-4fc9fb98879f",
    "description": "this is a second test",
    "name": "room_3",
    "roomId": "d10c80e8-49e4-4d6b-b966-4fc9fb98879f",
    "updatedAt": 1514851200000,
  },
]
`;

exports[`Rooms API /api/rooms POST should create a new room 1`] = `
Object {
  "createdAt": Any<Number>,
  "createdBy": "c10c80e8-49e4-4d6b-b966-4fc9fb98879f",
  "description": "a description",
  "name": "Living Room",
  "roomId": StringMatching /\\[a-fA-F0-9-\\]\\{36\\}/,
  "updatedAt": Any<Number>,
}
`;

exports[`Rooms API /api/rooms/:id GET should return an room 1`] = `
Object {
  "createdAt": 1514764800000,
  "createdBy": "c10c80e8-49e4-4d6b-b966-4fc9fb98879f",
  "description": "this is a test",
  "devices": Object {
    "dios": Array [
      Object {
        "dioId": 1,
        "name": "test_1",
        "roomId": "0fc1d78e-fd1c-4717-b610-65d2fa3d01b2",
      },
      Object {
        "dioId": 2,
        "name": "test_2",
        "roomId": "0fc1d78e-fd1c-4717-b610-65d2fa3d01b2",
      },
    ],
    "hueLights": Array [
      Object {
        "id": "1",
        "name": "light_1",
        "roomId": "0fc1d78e-fd1c-4717-b610-65d2fa3d01b2",
        "state": Object {
          "bri": 1,
          "on": true,
          "reachable": true,
        },
        "type": "Dimmable light",
      },
      Object {
        "id": "2",
        "name": "light_2",
        "roomId": "0fc1d78e-fd1c-4717-b610-65d2fa3d01b2",
        "state": Object {
          "bri": 1,
          "on": true,
          "reachable": true,
        },
        "type": "Dimmable light",
      },
    ],
  },
  "name": "room_1",
  "roomId": "0fc1d78e-fd1c-4717-b610-65d2fa3d01b2",
  "sensors": Array [
    Object {
      "id": "10",
      "state": Object {
        "lastupdated": "2020-05-09T19:10:00",
        "temperature": 20,
      },
      "type": "ZLLTemperature",
    },
  ],
  "updatedAt": 1514851200000,
}
`;

exports[`Rooms API /api/rooms/:id PATCH should update an room 1`] = `
Object {
  "createdAt": Any<Number>,
  "createdBy": "c10c80e8-49e4-4d6b-b966-4fc9fb98879f",
  "description": "this is a second test",
  "name": "room_updated",
  "roomId": StringMatching /\\[a-fA-F0-9-\\]\\{36\\}/,
  "updatedAt": Any<Number>,
}
`;
